feat(authz-keycloak): dynamic scope and resource mapping.

jenskeiner opened the pull request:
### What this PR does / why we need it:
Add code to `authz-keycloak` plugin to optionally resolve requests URI to resource(s) and map HTTP method to scope; fixes #3274.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

Open this PR to get some feedback on the changes. Still need to add tests to cover the new functionality. May still need to polish the JSON schema slightly as well. However, already tested changes in a real deployment to confirm this is actually working.

Also, strictly speaking, this plugin is not backwards compatible. E.g. we now require static permissions to be set when not using the new dynamic resource resolution. We also require `client_id` to be set because it will be used in any sensible configuration scenario. This mostly affects corner cases, so I don't think it represents a real problem.

In any case, will likely polish everything a bit more, but would appreciate any feedback already.

@sshniro You should definitely have a look.

jenskeiner commented on apisix/cli/ngx_tpl.lua:
```diff
@@ -371,16 +374,16 @@ http {
         {% end %}
         {% end %} {% -- if enable_ipv6 %}
 
+        {% if ssl.ssl_trusted_certificate ~= nil then %}
```
Moved this here since it was previously only used when APISIX itself was accepting HTTPS from the outside. But even if TLS is disabled, internally, plugins that send requests may still need to be able to use TLS and may have custom CA certs configured.

jenskeiner commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -42,7 +43,6 @@ local schema = {
             enum = {"urn:ietf:params:oauth:grant-type:uma-ticket"},
             minLength = 1, maxLength = 100
         },
-        audience = {type = "string", minLength = 1, maxLength = 100},
```
This is now called `client_id` which is more consistent with the `opendid-connect` plugin as well.

spacewander replied:
Maybe we can keep the old field and add a deprecated check?
https://github.com/apache/apisix/blob/bbbdf58d555ee89144b6923e751deabf7e0bbf15/apisix/init.lua#L463-L470

jenskeiner replied:
@spacewander Possible, it's just a name. So we keep `client_id` as well and prefer it, if given?

spacewander replied:
Yes. Maybe we can add a description on them:
https://github.com/apache/apisix/blob/bbbdf58d555ee89144b6923e751deabf7e0bbf15/apisix/schema_def.lua#L75

jenskeiner replied:
Ok, will change accordingly. Will also add/fix more test cases soon.

jenskeiner replied:
Ok, done.

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -224,31 +239,332 @@ local function authz_keycloak_get_token_endpoint(conf)
+
+                    authz_keycloak_cache_set("access_tokens",
+                                             token_endpoint .. ":" .. conf.client_id,
+                                             core.json.encode(session), 24 * 60 * 60)
```
Better to allow customizing cache time.

jenskeiner replied:
done

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -224,31 +239,332 @@ local function authz_keycloak_get_token_endpoint(conf)
+                    }
+                }
+
+                local current_time = ngx.time()
```
We already set `current_time` above?

jenskeiner replied:
Removed this invocation.

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -224,31 +239,332 @@ local function authz_keycloak_get_token_endpoint(conf)
+        end
+
+        authz_keycloak_cache_set("access_tokens", token_endpoint .. ":" .. conf.client_id,
+                                 core.json.encode(session), 24 * 60 * 60)
```
Ditto

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -224,31 +239,332 @@ local function authz_keycloak_get_token_endpoint(conf)
+-- computes access_token expires_in value (in seconds)
+local function authz_keycloak_access_token_expires_in(opts, expires_in)
+  return (expires_in or opts.access_token_expires_in or 300)
+         - 1 - (opts.access_token_expires_leeway or 0)
```
Where do we set `access_token_expires_in` and `access_token_expires_leeway`?

jenskeiner replied:
We don't set these at the moment; see next item below.

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -224,31 +239,332 @@ local function authz_keycloak_get_token_endpoint(conf)
+-- computes refresh_token expires_in value (in seconds)
+local function authz_keycloak_refresh_token_expires_in(opts, expires_in)
+  return (expires_in or opts.refresh_token_expires_in or 3600)
+         - 1 - (opts.refresh_token_expires_leeway or 0)
```
Ditto

jenskeiner replied:
This has been copied over from the `openidc` module. As far as I know, you cannot assume that the token endpoint will always return an `expires_in` and/or `refresh_expires_in` field as well. If it's not returned, this code allows to configure an expiry time assumption via `opts.refresh_token_expires_in`, falling back to a hard-coded default if that's not available neither.

Generally, I think having a fallback in place is useful. If needed, we could expose it through the schema as well. Same holds for the `leeway` parameter which can be used to make the plugin refresh tokens earlier. On the other hand, it's true that the mentioned options are currently not set anywhere.

Please advise whether this should be removed for now, or whether keeping it makes sense.

spacewander replied:
We could expose it through the schema. If you think it will make this PR more complex, you can remove it too.
It's all up to you.

jenskeiner replied:
Ok, will add it to the schema then. Better to have some levers to configure this in case of issues with token expiry.

spacewander commented on apisix/cli/ngx_tpl.lua:
```diff
@@ -146,6 +146,9 @@ http {
     lua_shared_dict introspection        10m; # cache for JWT verification results
 
+    # for authz-keycloak
+    lua_shared_dict access_tokens         1m; # cache for service account access tokens
```
Need to add it after:
https://github.com/apache/apisix/blob/bbbdf58d555ee89144b6923e751deabf7e0bbf15/t/APISIX.pm#L248

jenskeiner replied:
Will do. Can you quickly explain the reason?

spacewander replied:
The configuration used in test is generated from `apisix/t/APISIX.pm` instead of `apisix/cli/ngx_tpl.lua`.

jenskeiner replied:
Ah, ok.

jenskeiner replied:
@spacewander Quick question: We also need a shared dict for the discovery documents. I have added that to `apisix/t/APISIX.pm` now as well. But I can't see other dicts that e.g. the `openid-connect` plugin needs in that file. How do the tests then run successfully w/o the dict definitions?

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -53,10 +54,24 @@ local schema = {
+        lazy_load_paths = {type = "boolean", default = false},
+        http_method_as_scope = {type = "boolean", default = false},
     },
+    required = {"client_id"},
```
To avoid break change, we should not make `client_id` required. If `client_id` missed, we should continue the old logic.

jenskeiner replied:
done

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -53,10 +54,26 @@ local schema = {
+        {required = {"token_endpoint"}}
+    },
+    dependencies = {
+        lazy_load_paths = {
```
Should we not give a default value to `lazy_load_paths`?
Currently either discovery or resource_registration_endpoint is required as there is always a value for lazy_load_paths.

jenskeiner replied:
Fixing. Defaults to `false`. Should require one of `discovery` or `resource_registration_endpoint` if set to `true`.

jenskeiner replied:
Should be fixed now.

jenskeiner commented:
I reckon the new tests added are sufficient to check the basic validity of the updated schema wrt. to (conditionally) required attributes like `resource_registration_endpoint`.

However, it would make sense to add more cases to test the actual functionality of `lazy_load_paths` and `http_method_as_scope`. The problem is that this would very likely require some modifications to the Keycloak container image since we need resources linked to scopes that correspond to HTTP method names. Also would probably need to verify resource definitions' URI settings so we can test the resource resolution via the resource registration endpoint.

Feedback?

The Docker image is in @sshniro's private repo. Can you advise @sshniro how we should go about making modifications?

jenskeiner commented:
Also, I could use some help on test case #7. I want to conditionally require either `discovery` or `resource_registration_endpoint` when `lazy_load_paths` is set to `true`. But there must be something wrong with my schema definition where I'm currently trying to use `if` and `then` (part of newer schema draft standards). The invalid configuration in case #7 seems to be accepted -- which is wrong. Can anybody help? Maybe @spacewander?

spacewander commented:
The `if` and `then` schema is not supported by the Lua implementation of JSONSchema yet.

jenskeiner commented:
> The `if` and `then` schema is not supported by the Lua implementation of JSONSchema yet.

Ok, I think I'm on top of it now. Just used `anyOf` to the same effect. Minor point is that the error message that the schema validator returns  -- see test case 7 -- is not really detailed. But it works as expected.

jenskeiner commented:
I've switched to an updated Keycloak image in my own DockerHub repo. This one has some slight configuration changes to be able to test the URI-to-resource and HTTP-method-to-scope mapping.

Also, other tests currently disabled to make the checks shorter. Will re-enable when I've added more test cases.

jenskeiner commented:
When @sshniro has given me access to his Docker repo, I will update the Keycloak image there and revert the change in the test runner script to pull the image from a different repo.

sshniro commented:
LGTM, also @jenskeiner I have granted access to the repo, would be great if we can tag this image, rather than replacing the latest.

jenskeiner commented:
> LGTM, also @jenskeiner I have granted access to the repo, would be great if we can tag this image, rather than replacing the latest.

Ok, great. Any preference as to which tag name I should use?

spacewander commented on doc/plugins/authz-keycloak.md:
```diff
@@ -38,24 +38,37 @@ For more information on Keycloak, refer to [Keycloak Authorization Docs](https:/
+| token_endpoint                 | string        | optional    |                                               | https://host.domain/auth/realms/foo/protocol/openid-connect/token  | A OAuth2-compliant Token Endpoint that supports the `urn:ietf:params:oauth:grant-type:uma-ticket` grant type. Overrides value from discovery, if given.     |
+| resource_registration_endpoint | string        | optional    |                                               | https://host.domain/auth/realms/foo/authz/protection/resource_set  | A Keycloak Protection API-compliant resource registration endpoint. Overrides value from discovery, if given.                                               |
+| grant_type                     | string        | optional    | "urn:ietf:params:oauth:grant-type:uma-ticket" | ["urn:ietf:params:oauth:grant-type:uma-ticket"]                    |                                                                                                                                                             |
+| client_id                      | string        | required    |                                               |                                                                    | The client identifier of the resource server to which the client is seeking access. <br>This parameter is mandatory when parameter permission is defined.   |
```
> This parameter is mandatory when parameter permission is defined.

Look like it should be `when parameter lazy_load_paths is defined`? Should require it in the schema and add a test for it.

jenskeiner replied:
Ok, so I'm now again requiring one of `client_id` or `audience`, the latter for better backwards compatibility. Rationale: The plugin always uses the token endpoint to get a decision from Keycloak on the requested permissions. As per https://www.keycloak.org/docs/4.8/authorization_services/#_service_obtaining_permissions the `audience` parameter (what is either in `client_id` or `audience` for the plugin) is required when the `permission` parameter is set (which the plugin always does).

If `lazy_load_paths` is `true`, the the plugin also needs to obtain an access token for itself from Keycloak. To that end, it most likely also needs `client_secret` to be set, but that depends on how Keycloak is configured. Hence, this is still optional.

Also added test cases to check if one of `client_id` or `audience` is correctly required and schema verification fails if both are absent.

spacewander replied:
I see.

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -224,31 +275,326 @@ local function authz_keycloak_get_token_endpoint(conf)
+    end
+
+    local session = authz_keycloak_cache_get("access_tokens", token_endpoint .. ":"
+                                             .. client_id)
```
The `client_id` could be nil if you don't force when lazy_load_paths is used.

jenskeiner replied:
See above, we're now requiring `client_id` (or `audience`).

spacewander commented on t/plugin/authz-keycloak.t:
```diff
@@ -137,7 +224,7 @@ done
                             "authz-keycloak": {
                                 "token_endpoint": "http://127.0.0.1:8090/auth/realms/University/protocol/openid-connect/token",
                                 "permissions": ["course_resource#view"],
-                                "audience": "course_management",
```
Maybe we can leave `audience` unchanged in one or two cases so that we know the old configuration is still compatible?

jenskeiner replied:
Sure, I can add some cases that use it.

spacewander commented:
BTW, you can split the test into two files, like the https://github.com/apache/apisix/blob/master/t/plugin/limit-conn2.t. Since it is too big now.

jenskeiner commented on doc/plugins/authz-keycloak.md:
```diff
@@ -38,24 +38,37 @@ For more information on Keycloak, refer to [Keycloak Authorization Docs](https:/
+| policy_enforcement_mode        | string        | optional    | "ENFORCING"                                   | ["ENFORCING", "PERMISSIVE"]                                        |                                                                                                                                                             |
+| permissions                    | array[string] | optional    |                                               |                                                                    | Static permission to request, an array of strings each representing a resources and optionally one or more scopes the client is seeking access.             |
+| lazy_load_paths                | boolean       | optional    | false                                         |                                                                    | Dynamically resolve the request URI to resource(s) using the resource registration endpoint instead of using the static permission.                         |
+| http_method_as_scope           | boolean       | optional    | false                                         |                                                                    | Map HTTP request type to scope of same name and add to all permissions requested.                                                                           |
```
Correct. Will add.

jenskeiner replied:
Added.

jenskeiner commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -171,13 +222,13 @@ local function authz_keycloak_discover(url, ssl_verify, keepalive, timeout,
       keepalive = (keepalive ~= "no")
```
My bad, was copied over from `openidc` module. Fixed it and consolidated the HTTP client setup -- which is done in various places -- to apply options like `timeout`, `keepalive`, `ssl_verify`and others consistently.

jenskeiner commented:
> BTW, you can split the test into two files, like the https://github.com/apache/apisix/blob/master/t/plugin/limit-conn2.t. Since it is too big now.

done

jenskeiner commented:
@spacewander I think I've addressed all of the feedback so far. Thanks for reviewing. Let me know any further issues.

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -69,225 +109,566 @@ local _M = {
+-- Return access_token expires_in value (in seconds).
+local function authz_keycloak_access_token_expires_in(conf, expires_in)
+    return (expires_in or conf.access_token_expires_in or 300)
+           - 1 - (conf.access_token_expires_leeway or 0)
```
Will you add this configurations in this PR, or in the future one?

spacewander commented on apisix/plugins/authz-keycloak.lua:
```diff
@@ -69,225 +109,566 @@ local _M = {
+-- Return refresh_token expires_in value (in seconds).
+local function authz_keycloak_refresh_token_expires_in(conf, expires_in)
+    return (expires_in or conf.refresh_token_expires_in or 3600)
+           - 1 - (conf.refresh_token_expires_leeway or 0)
```
Ditto

spacewander reviewed (approved):
We can merge this PR first and solve the remain in the next one.

jenskeiner commented:
@spacewander Thanks, I will surely continue the work on this plugin and I can add the config options in a future PR.

membphis commented:
I have been busy recently, I will review this PR later.

Welcome everyone to review.  many thx
